src/lib/GenericPredictor.cc

```diff
@@ -115,6 +127,20 @@ void GenericPredictor::flush(uint64_t address) {
     }
     rasHistory_.erase(it);
   }
+
+  // If possible, pop instruction from FTQ
+  FTQ_.pop_back();
+
+  // Roll back global history
+  globalHistory_ >>= 1;
+}
+
+void GenericPredictor::addToFTQ(uint64_t address, bool taken) {
+  // Make the hashed index and add it to the FTQ
+  uint64_t hashedIndex = (address ^ globalHistory_) & ((1 << btbBits_) - 1);
+  FTQ_.emplace_back(taken, hashedIndex);
+  // Speculatively update the global history
+  globalHistory_ = ((globalHistory_ << 1) | taken) & globalHistoryLength_;
```

Comment:
Could there ever be a scenario where `addToFTQ` and `predict` are called for the same branch instruction; adding said branch to the FTQ and globalHistory twice?

Reply:
No, they are called by mutually exclusive conditional routes through the tick() function in the fetch unit.  addToFTQ() is called in the event that (loopBufferState_ == LoopBufferState::SUPPLYING), and predict() is called otherwise.